### PR TITLE
mrc-1758: New wrapper class for C models

### PR DESCRIPTION
**Pull request by richfitz**
This PR removes the old way of creating classes by building methods on the fly and injecting them into a class-constructor in the package, by instead using the same approach as `dust`. Now, we keep a template within odin and implement separate `run` functions. A standalone version is written into the destination package, which simplifies things. This also allows for C symbol registration (fixing #153) and eliminated byte compilation warnings

<details>
<summary>squire</summary>

```
rfitzjoh@wpia-dide300:~/Documents/Projects/ncov/squire$ R CMD INSTALL .
* installing to library ‘/home/rfitzjoh/lib/R/library’
* installing *source* package ‘squire’ ...
** using staged installation
** libs
make: Nothing to be done for 'all'.
installing to /home/rfitzjoh/lib/R/library/00LOCK-squire/00new/squire/libs
** R
** data
*** moving datasets to lazyload DB
** inst
** byte-compile and prepare package for lazy loading
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
Error: compilation failed -  the empty environment has no parent at parent.env(env) 
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (squire)
```

</details>

This adds a number of breaking changes (see NEWS) and I've put off further work for subsequent PRs:

* Fix use of use_dde by moving to run, rather than constructor
* Deprecate use of construtor function
* Deprecate use of deriv/update directly
* Add docs to the interface and generate that into package (like in dust)
* Port to roxygen + markdown
* Check that the R and C interfaces align
* Port R interface to use a fixed class and helpers (not needed right now, ever?)

### CODE REVIEW

File: R/wrapper.R [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    colnames(ret) <- NULL
+  }
+
+  ret
+}
+
+
+wrapper_run_delay <- function(self, private, t, y, ...,
+                              use_names = TRUE, tcrit = NULL,
+                              n_history = DEFAULT_HISTORY_SIZE) {
```
Comment (weshinsley) [inline R/wrapper.R:136]: Is there any risk of problems with the position of "..." here?

File: R/wrapper.R [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    colnames(ret) <- private$ynames
+  } else {
+    colnames(ret) <- NULL
+  }
+
+  ret
+}
+
+wrapper_run_discrete <- function(self, private, step, y = NULL, ...,
+                                 use_names = TRUE, replicate = NULL) {
```
Comment (weshinsley) [inline R/wrapper.R:176]: Ditto - does '...' need to be last?

File: R/wrapper.R (hunk 1 above)
Comment (richfitz) [reply R/wrapper.R:136]: `...` at that point just means the following args **must** be passed by name (which is the old interface too)

**Review by weshinsley — APPROVED** [review]
This looks good as far as I can tell - I will try and get more familiar with odin over time...

**Pull request merged**